src/transformers/index.ts

```diff
@@ -0,0 +1,179 @@
+import {
+  SchemaObject,
+  ReferenceObject,
+  MapSubSchema,
+  ZodNumericCheck,
+} from '../types';
+import { ZodType } from 'zod';
+import { UnknownZodTypeError } from '../errors';
+import { isZodType } from '../lib/zod-is-type';
+import { Metadata } from '../metadata';
+import { ArrayTransformer } from './array';
+import { BigIntTransformer } from './big-int';
+import { DiscriminatedUnionTransformer } from './discriminated-union';
+import { EnumTransformer } from './enum';
+import { IntersectionTransformer } from './intersection';
+import { LiteralTransformer } from './literal';
+import { NativeEnumTransformer } from './native-enum';
+import { NumberTransformer } from './number';
+import { ObjectTransformer } from './object';
+import { RecordTransformer } from './record';
+import { StringTransformer } from './string';
+import { TupleTransformer } from './tuple';
+import { UnionTransformer } from './union';
+import { OpenApiVersionSpecifics } from '../openapi-generator';
+
+export class OpenApiTransformer {
+  constructor(private versionSpecifics: OpenApiVersionSpecifics) {}
+
+  transform<T>(
+    zodSchema: ZodType<T>,
+    isNullable: boolean,
+    mapItem: MapSubSchema,
+    generateSchemaRef: (ref: string) => string,
+    defaultValue?: T
+  ): SchemaObject | ReferenceObject {
+    if (isZodType(zodSchema, 'ZodNull')) {
+      return this.versionSpecifics.nullType;
+    }
+
+    if (isZodType(zodSchema, 'ZodUnknown') || isZodType(zodSchema, 'ZodAny')) {
+      return this.versionSpecifics.mapNullableType(undefined, isNullable);
+    }
+
+    if (isZodType(zodSchema, 'ZodObject')) {
+      return new ObjectTransformer().transform(
+        zodSchema,
+        defaultValue as object, // verified on TS level from input
+        _ => this.versionSpecifics.mapNullableType(_, isNullable),
+        mapItem
+      );
+    }
+
+    const schema = this.transformSchemaWithoutDefault(
+      zodSchema,
+      isNullable,
+      mapItem,
+      generateSchemaRef
+    );
+
+    return { ...schema, default: defaultValue };
+  }
+
+  private transformSchemaWithoutDefault<T>(
+    zodSchema: ZodType<T>,
+    isNullable: boolean,
+    mapItem: MapSubSchema,
+    generateSchemaRef: (ref: string) => string
+  ): SchemaObject | ReferenceObject {
+    if (isZodType(zodSchema, 'ZodUnknown') || isZodType(zodSchema, 'ZodAny')) {
+      return this.versionSpecifics.mapNullableType(undefined, isNullable);
+    }
+
+    if (isZodType(zodSchema, 'ZodString')) {
+      return new StringTransformer().transform(zodSchema, schema =>
+        this.versionSpecifics.mapNullableType(schema, isNullable)
+      );
+    }
+
+    if (isZodType(zodSchema, 'ZodNumber')) {
+      return new NumberTransformer().transform(
```

Comment:
Can we extract these "new" instances as private fields, so that we don't need to create the object for every transform? Not sure how much performance difference there is but seems better to be on the safe side since this is a hot path